File: backend/app/schemas/course_design.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, model_validator
from pydantic.alias_generators import to_camel


class CourseDesignSchema(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True, serialize_by_alias=True, str_strip_whitespace=True)


class CourseDesignProject(CourseDesignSchema):
    project_id: int
    title: str
    topic: str
    project_type: str


class CourseDesignContext(CourseDesignSchema):
    grade: int | None = None
    class_hours: int | None = None
    lesson_minutes: int | None = None
    class_size: int | None = None
    student_level: str | None = None
    student_experience: str | None = None
    ai_access_mode: str | None = None
    device_condition: str | None = None
    devices: list[Any] = Field(default_factory=list)
    constraints: list[Any] = Field(default_factory=list)
    additional_requirements: str | None = None


class CourseContextDiagnosis(CourseDesignSchema):
    core_problem: str = Field(min_length=1)
    existing_foundation: str = Field(min_length=1)
    learning_difficulties: list[str] = Field(min_length=1)
    constraints: list[str] = Field(min_length=1)


class CourseContextDiagnosisRequest(CourseDesignSchema):
    project_id: int = Field(gt=0)
    grade: int = Field(gt=0)
    topic: str = Field(min_length=1)
    lesson_minutes: int = Field(gt=0)
    class_size: int = Field(gt=0)
    student_experience: str = Field(min_length=1)
    device_condition: str = Field(min_length=1)
    additional_requirements: str = ""
    prompt: str = Field(min_length=1)


class CourseObjectiveProposal(CourseDesignSchema):
    content: str = Field(min_length=1)
    rationale: str = Field(min_length=1)
    standard_refs: list[int] = Field(default_factory=list)
    literacy_refs: list[int] = Field(default_factory=list)


class CourseObjectiveGenerationRequest(CourseDesignSchema):
    project_id: int = Field(gt=0)
    grade: int = Field(gt=0)
    topic: str = Field(min_length=1)
    context_diagnosis: CourseContextDiagnosis
    evidence: list[dict[str, str]] = Field(default_factory=list)
    curriculum_standards: list[dict[str, object]] = Field(default_factory=list)
    ai_literacy_items: list[dict[str, object]] = Field(default_factory=list)
    prompt: str = Field(min_length=1)


class CourseObjectiveGenerationResult(CourseDesignSchema):
    objectives: list[CourseObjectiveProposal] = Field(min_length=2, max_length=4)


class CourseObjectiveUpdateRequest(CourseDesignSchema):
    content: str = Field(min_length=1)
    action: str = Field(pattern="^REVISE$")


class CourseObjectiveCreateRequest(CourseDesignSchema):
    content: str = Field(min_length=1)


class CoursePedagogyRecommendation(CourseDesignSchema):
    method_id: int = Field(gt=0)
    name: str = Field(min_length=1)
    rationale: str = Field(min_length=1)
    components: list[str] = Field(min_length=1)


class CoursePedagogyRecommendationRequest(CourseDesignSchema):
    project_id: int = Field(gt=0)
    context: dict[str, object]
    objectives: list[dict[str, object]] = Field(min_length=1)
    evidence: list[dict[str, str]] = Field(default_factory=list)
    methods: list[dict[str, object]] = Field(min_length=1)
    prompt: str = Field(min_length=1)


class CoursePedagogyRecommendationResult(CourseDesignSchema):
    recommended: CoursePedagogyRecommendation
    alternatives: list[CoursePedagogyRecommendation] = Field(default_factory=list)


class CoursePedagogySelectRequest(CourseDesignSchema):
    method_id: int = Field(gt=0)


class CoursePedagogyCustomRequest(CourseDesignSchema):
    name: str = Field(min_length=1, max_length=255)
    description: str = Field(min_length=1)


class CourseAssessmentProposal(CourseDesignSchema):
    objective_id: int = Field(gt=0)
    task_content: str = Field(min_length=1)
    student_evidence: list[str] = Field(min_length=1)
    criteria: list[str] = Field(min_length=1)
    rationale: str = Field(min_length=1)


class CourseAssessmentGenerationRequest(CourseDesignSchema):
    project_id: int = Field(gt=0)
    context: dict[str, object]
    objectives: list[dict[str, object]] = Field(min_length=1)
    pedagogy: dict[str, object]
    ai_literacy_items: list[dict[str, object]] = Field(default_factory=list)
    evidence: list[dict[str, str]] = Field(default_factory=list)
    prompt: str = Field(min_length=1)


class CourseAssessmentGenerationResult(CourseDesignSchema):
    assessments: list[CourseAssessmentProposal] = Field(min_length=1)


class CourseAssessmentUpdateRequest(CourseDesignSchema):
    task_content: str | None = Field(default=None, min_length=1)
    student_evidence: list[str] | None = None
    criteria: list[str] | None = None

    @model_validator(mode="after")
    def require_a_change(self):
        if self.task_content is None and self.student_evidence is None and self.criteria is None:
            raise ValueError("At least one assessment field must be supplied")
        return self


class CourseActivityProposal(CourseDesignSchema):
    name: str = Field(min_length=1, max_length=255)
    duration: int = Field(gt=0)
    core_task: str = Field(min_length=1)
    teacher_action: str = Field(min_length=1)
    student_action: str = Field(min_length=1)
    ai_role: str = Field(min_length=1, max_length=64)
    assessment: str = Field(min_length=1)
    scaffolds: list[str] = Field(default_factory=list)
    objective_refs: list[int] = Field(default_factory=list)


class CourseBlueprintGenerationRequest(CourseDesignSchema):
    project_id: int = Field(gt=0)
    context: dict[str, object]
    objectives: list[dict[str, object]] = Field(min_length=1)
    pedagogy: dict[str, object]
    assessments: list[dict[str, object]] = Field(min_length=1)
    evidence: list[dict[str, str]] = Field(default_factory=list)
    prompt: str = Field(min_length=1)


class CourseBlueprintGenerationResult(CourseDesignSchema):
    activities: list[CourseActivityProposal] = Field(min_length=4, max_length=4)


class CourseActivityRegenerationRequest(CourseDesignSchema):
    blueprint: CourseBlueprintGenerationRequest
    activity: CourseActivityProposal
    previous_activity: CourseActivityProposal | None = None
    next_activity: CourseActivityProposal | None = None


class CourseActivityTransformRequest(CourseDesignSchema):
    action: str = Field(pattern="^(SHORTEN|INCREASE_DIFFICULTY|DECREASE_DIFFICULTY|ADD_SCAFFOLD)$")


class CourseActivityUpdateRequest(CourseDesignSchema):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    duration: int | None = Field(default=None, gt=0)
    core_task: str | None = Field(default=None, min_length=1)
    teacher_action: str | None = Field(default=None, min_length=1)
    student_action: str | None = Field(default=None, min_length=1)
    ai_role: str | None = Field(default=None, min_length=1, max_length=64)
    assessment: str | None = Field(default=None, min_length=1)
    scaffolds: list[str] | None = None

    @model_validator(mode="after")
    def require_a_change(self):
        if all(
            value is None
            for value in (
                self.name, self.duration, self.core_task, self.teacher_action,
                self.student_action, self.ai_role, self.assessment, self.scaffolds,
            )
        ):
            raise ValueError("At least one activity field must be supplied")
        return self


class CourseQualityCheckProposal(CourseDesignSchema):
    check_type: str = Field(min_length=1, max_length=64)
    status: str = Field(pattern="^(PASS|WARNING)$")
    issue: str | None = None
    reason: str | None = None
    suggestion: str | None = None
    evidence: dict[str, Any] = Field(default_factory=dict)


class CourseQualityCheckRequest(CourseDesignSchema):
    project_id: int = Field(gt=0)
    context: dict[str, object]
    objectives: list[dict[str, object]] = Field(min_length=1)
    pedagogy: dict[str, object]
    assessments: list[dict[str, object]] = Field(min_length=1)
    activities: list[dict[str, object]] = Field(min_length=1)
    evidence: list[dict[str, str]] = Field(default_factory=list)
    prompt: str = Field(min_length=1)


class CourseQualityCheckResult(CourseDesignSchema):
    checks: list[CourseQualityCheckProposal] = Field(min_length=1)


class CourseDesignObjective(CourseDesignSchema):
    objective_id: int
    content: str
    sequence_no: int
    standard_refs: list[Any] = Field(default_factory=list)
    literacy_refs: list[Any] = Field(default_factory=list)
    rationale: str | None = None
    source_type: str
    teacher_action: str | None = None
    confirmed: bool
    version: int


class CourseDesignPedagogy(CourseDesignSchema):
    pedagogy_id: int | None = None
    primary_method_id: int | None = None
    secondary_method_id: int | None = None
    rationale: str | None = None
    suitable_for: list[Any] = Field(default_factory=list)
    risk_note: str | None = None
    alternatives: dict[str, Any] | list[Any] | None = None
    teacher_action: str | None = None
    confirmed: bool | None = None
    version: int | None = None
    custom_name: str | None = None
    custom_description: str | None = None
    source_type: str | None = None


class CourseDesignAssessment(CourseDesignSchema):
    assessment_id: int
    objective_id: int
    task_content: str
    student_evidence: list[Any] = Field(default_factory=list)
    criteria: dict[str, Any] | list[Any] = Field(default_factory=list)
    rationale: str | None = None
    teacher_action: str | None = None
    confirmed: bool
    version: int


class CourseDesignActivity(CourseDesignSchema):
    activity_id: int
    sequence_no: int
    name: str
    duration: int
    core_task: str | None = None
    teacher_action: str | None = None
    student_action: str | None = None
    ai_role: str
    dominant_actor: str
    assessment_note: str | None = None
    scaffolds: list[Any] = Field(default_factory=list)
    objective_refs: list[Any] = Field(default_factory=list)
    version: int


class CourseDesignQualityCheck(CourseDesignSchema):
    quality_check_id: int
    check_type: str
    status: str
    issue: str | None = None
    reason: str | None = None
    suggestion: str | None = None
    evidence: dict[str, Any] | list[Any] | None = None
    created_at: datetime


class CourseDesignState(CourseDesignSchema):
    project: CourseDesignProject
    context: CourseDesignContext
    context_diagnosis: dict[str, Any] = Field(default_factory=dict)
    objectives: list[CourseDesignObjective] = Field(default_factory=list)
    pedagogy: CourseDesignPedagogy = Field(default_factory=CourseDesignPedagogy)
    assessments: list[CourseDesignAssessment] = Field(default_factory=list)
    activities: list[CourseDesignActivity] = Field(default_factory=list)
    quality_checks: list[CourseDesignQualityCheck] = Field(default_factory=list)
    workflow_state: str
    stale_sections: list[str] = Field(default_factory=list)
    current_step: int
    completed_steps: list[int] = Field(default_factory=list)

File: frontend/src/api/courseDesign.ts
import http from './http'

export const AI_REQUEST_TIMEOUT = 260_000
const aiRequest = { timeout: AI_REQUEST_TIMEOUT }

export interface ApiEnvelope<T> { code: number; message: string; data: T; requestId: string }
export interface CourseObjective { objectiveId: number; content: string; sourceType: string; teacherAction?: string | null; confirmed: boolean; version: number }
export interface CourseAssessment { assessmentId: number; objectiveId: number; taskContent: string; studentEvidence: string[]; criteria: string[]; teacherAction?: string | null; confirmed: boolean; version: number }
export interface CourseActivity { activityId: number; sequenceNo: number; name: string; duration: number; coreTask: string; teacherAction: string; studentAction: string; aiRole: string; assessment?: string; assessmentNote?: string; scaffolds: string[]; objectiveRefs: number[]; version: number }
export interface CourseDesignState {
  project: { projectId: number; title: string; topic: string }
  context: { grade?: number; lessonMinutes?: number; classSize?: number; studentExperience?: string; deviceCondition?: string; devices?: string[]; additionalRequirements?: string }
  contextDiagnosis: { coreProblem: string; existingFoundation: string; learningDifficulties: string[]; constraints: string[] }
  objectives: CourseObjective[]
  pedagogy: Record<string, any>
  assessments: CourseAssessment[]
  activities: CourseActivity[]
  qualityChecks: QualityCheck[]
  workflowState: string
  staleSections: string[]
  currentStep: number
  completedSteps: number[]
}
export interface QualityCheck { qualityCheckId: number; checkType: string; status: string; issue?: string; reason?: string; suggestion?: string; evidence?: Record<string, any> }

const projectPath = (projectId: number | string) => '/projects/' + projectId
export const getCourseDesign = (projectId: number | string) => http.get<ApiEnvelope<CourseDesignState>>(projectPath(projectId) + '/course-design')
export const saveContext = (projectId: number | string, payload: Record<string, any>) => http.put(projectPath(projectId) + '/context', payload)
export const diagnoseContext = (projectId: number | string) => http.post(projectPath(projectId) + '/context/diagnose', undefined, aiRequest)
export const confirmContext = (projectId: number | string) => http.post(projectPath(projectId) + '/context/confirm')
export const generateObjectives = (projectId: number | string) => http.post(projectPath(projectId) + '/objectives/generate', undefined, aiRequest)
export const keepObjective = (projectId: number | string, objectiveId: number) => http.post(projectPath(projectId) + '/objectives/' + objectiveId + '/keep')
export const updateObjective = (projectId: number | string, objectiveId: number, content: string) => http.patch(projectPath(projectId) + '/objectives/' + objectiveId, { content, action: 'REVISE' })
export const deleteObjective = (projectId: number | string, objectiveId: number) => http.delete(projectPath(projectId) + '/objectives/' + objectiveId)
export const addObjective = (projectId: number | string, content: string) => http.post(projectPath(projectId) + '/objectives', { content })
export const confirmObjectives = (projectId: number | string) => http.post(projectPath(projectId) + '/objectives/confirm')
export const recommendPedagogy = (projectId: number | string) => http.post(projectPath(projectId) + '/pedagogy/recommend', undefined, aiRequest)
export const selectPedagogy = (projectId: number | string, methodId: number) => http.post(projectPath(projectId) + '/pedagogy/select', { methodId })
export const createCustomPedagogy = (projectId: number | string, name: string, description: string) => http.post(projectPath(projectId) + '/pedagogy/custom', { name, description })
export const confirmPedagogy = (projectId: number | string) => http.post(projectPath(projectId) + '/pedagogy/confirm')
export const generateAssessments = (projectId: number | string) => http.post(projectPath(projectId) + '/assessments/generate', undefined, aiRequest)
export const updateAssessment = (projectId: number | string, assessmentId: number, payload: Record<string, any>) => http.patch(projectPath(projectId) + '/assessments/' + assessmentId, payload)
export const regenerateAssessment = (projectId: number | string, assessmentId: number) => http.post(projectPath(projectId) + '/assessments/' + assessmentId + '/regenerate', undefined, aiRequest)
export const confirmAssessments = (projectId: number | string) => http.post(projectPath(projectId) + '/assessments/confirm')
export const generateBlueprint = (projectId: number | string, lessonMinutes: number) =>
  http.post(projectPath(projectId) + '/course-blueprint/generate', { lessonMinutes }, aiRequest)
export const updateActivity = (projectId: number | string, activityId: number, payload: Record<string, any>) => http.patch(projectPath(projectId) + '/activities/' + activityId, payload)
export const regenerateActivity = (projectId: number | string, activityId: number) => http.post(projectPath(projectId) + '/activities/' + activityId + '/regenerate', undefined, aiRequest)
export const transformActivity = (projectId: number | string, activityId: number, action: string) => http.post(projectPath(projectId) + '/activities/' + activityId + '/transform', { action })
export const getDesignEvidence = (projectId: number | string, bizType: string, bizId: number) => http.get(projectPath(projectId) + '/design-evidence', { params: { bizType, bizId } })
export const runQualityCheck = (projectId: number | string) => http.post(projectPath(projectId) + '/quality-check', undefined, aiRequest)
export const getQualityCheck = (projectId: number | string) => http.get(projectPath(projectId) + '/quality-check')
export const applyQualitySuggestion = (projectId: number | string, checkId: number) => http.post(projectPath(projectId) + '/quality-check/' + checkId + '/apply')
